class Solution {
    public int climbStairs(int n) {
        int[] memo = new int[n + 1];
        return ways(n, memo);
    }

    private int ways(int n, int[] memo) {
        if (n == 0 || n == 1) {
            return 1;
        }

        if (memo[n] != 0) {
            return memo[n];
        }

        memo[n] = ways(n - 1, memo) + ways(n - 2, memo);

        return memo[n];
    }
}
